src/realm/object-store/c_api/object.cpp

```diff
@@ -0,0 +1,540 @@
+        if (pk_val.is_null() && !pk_col.is_nullable()) {
+            if (out_found)
+                *out_found = false;
+            return nullptr;
+        }
+        if (!pk_val.is_null() && ColumnType(pk_val.get_type()) != pk_col.get_type() &&
+            pk_col.get_type() != col_type_Mixed) {
+            if (out_found)
+                *out_found = false;
+            return nullptr;
+        }
+
+        auto obj_key = table->find_primary_key(pk_val);
+        if (obj_key) {
+            if (out_found)
+                *out_found = true;
+            auto obj = table->get_object(obj_key);
+            return new realm_object_t{Object{shared_realm, std::move(obj)}};
+        }
+        else {
+            if (out_found)
+                *out_found = false;
+            return static_cast<realm_object_t*>(nullptr);
+        }
+    });
+}
+
+RLM_API realm_object_t* realm_object_create(realm_t* realm, realm_table_key_t table_key)
+{
+    return wrap_err([&]() {
+        auto& shared_realm = *realm;
+        auto tblkey = from_capi(table_key);
+        auto table = shared_realm->read_group().get_table(tblkey);
+
+        if (table->get_primary_key_column()) {
+            auto& object_schema = schema_for_table(realm, table_key);
+            throw MissingPrimaryKeyException{object_schema.name};
+        }
+
+        auto obj = table->create_object();
+        auto object = Object{shared_realm, std::move(obj)};
+        return new realm_object_t{std::move(object)};
+    });
+}
+
+RLM_API realm_object_t* realm_object_create_with_primary_key(realm_t* realm, realm_table_key_t table_key,
+                                                             realm_value_t pk)
+{
+    return wrap_err([&]() {
+        auto& shared_realm = *realm;
+        auto tblkey = from_capi(table_key);
+        auto table = shared_realm->read_group().get_table(tblkey);
+        // FIXME: Provide did_create?
+        auto pkval = from_capi(pk);
+
+        ColKey pkcol = table->get_primary_key_column();
+        if (!pkcol) {
+            // FIXME: Proper exception type.
+            throw std::logic_error("Class does not have a primary key");
+        }
+
+        if (pkval.is_null() && !pkcol.is_nullable()) {
+            auto& schema = schema_for_table(realm, table_key);
+            throw NotNullableException{schema.name, schema.primary_key};
+        }
+
+        if (!pkval.is_null() && pkval.get_type() != DataType(pkcol.get_type())) {
+            auto& schema = schema_for_table(realm, table_key);
+            throw WrongPrimaryKeyTypeException{schema.name};
+        }
+
+        auto obj = table->create_object_with_primary_key(pkval);
+        auto object = Object{shared_realm, std::move(obj)};
+        return new realm_object_t{std::move(object)};
+    });
+}
+
+RLM_API bool realm_object_delete(realm_object_t* obj)
+{
+    return wrap_err([&]() {
+        auto o = obj->obj();
+
+        if (!obj->is_valid()) {
```

Comment:
it's cheaper to try a remove and catch a failure than to check separately with is_valid.